Use explicit tenant UUID returned by tenant

gkrenn opened the pull request:
<!--

Thanks for opening a pull request! Here are some tips to get everything merged smoothly:

1. Read our contributor guidelines: https://github.com/Dynatrace/dynatrace-operator/blob/main/CONTRIBUTING.md

2. If the PR is unfinished, raise it as a draft or prefix the title with "WIP:" so it's clear to everyone.

3. Be sure to allow edits from maintainers, so it's easier for us to help: https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

-->

## Description

Currently the tenant UUID is derived from the URL provided in the dynakube.
It is never guaranteed that the url contains a valid UUID and can even be an alias. 

This PR updates the corresponding code to use the UUID returned by the tenant itself.

[Jira Ticket](https://dt-rnd.atlassian.net/browse/K8S-9537)
<!--

Please include the following:
- The motivation for the change
    - Link to the Github issue or Jira ticket, if exists.
- The summary of the change

-->

## How can this be tested?

Run cloudNative e2e test and check if everything works as expected. 
Run resilience test to verify that even with limited connectivity the CSI driver works as expected.

<!--

Please include some guiding steps on how to test this change during a review.

- What environment is necessary for the change to be noticeable ?
- What needs to be enabled/applied for the change to be noticeable ?

-->

0sewa0 commented on pkg/api/v1beta1/dynakube/properties.go:
```diff
@@ -414,9 +414,14 @@ func (dk *DynaKube) Tokens() string {
-// TenantUUIDFromApiUrl gets the tenantUUID from the ApiUrl present in the struct, if the tenant is aliased then the alias will be returned.
-func (dk *DynaKube) TenantUUIDFromApiUrl() (string, error) {
-	return tenantUUID(dk.Spec.APIURL)
+func (dk *DynaKube) TenantUUIDFromConnectionInfo() (string, error) {
```
```suggestion
func (dk *DynaKube) TenantUUID() (string, error) {
```

as there is no other `TennantUUIDFromXYZ` function, just having it with a shorter name would be better

0sewa0 commented on pkg/controllers/dynakube/controller.go:
```diff
@@ -297,23 +301,15 @@ func (controller *Controller) setupTokensAndClient(ctx context.Context, dynakube
 func (controller *Controller) reconcileComponents(ctx context.Context, dynatraceClient dtclient.Client, istioClient *istio.Client, dynakube *dynatracev1beta1.DynaKube) error {
 	var componentErrors []error
 
+	pullSecretReconciler := dtpullsecret.
+		NewReconciler(controller.client, controller.apiReader, dynakube, controller.tokens)
+
```
why is this change necessary for the `using the tennantUUID from the connection-info` logic?

Also, it looks kinda wrong (sorry if its a bit rambly 😅 )
- when do you **not** need a pull-secret? => when its application-monitoring
   - for both oneagent and activegate it is needed
   - so why move it into the `injection` reconciler?
   - why move it at all in the context of the intended change?
 - why store a "status" in memory that will be released right after each reconcile?
   - The current way we do it for the`connectionInfo` is that the `injection` and `oneagent` Reconcilers don't each get a `connection-info-reconciler` but create one on there own, without coupling everything together.
      - The "is already reconciled" logic is done using the status and the timestamp within it, so its multipurpose
         - Only gets updated every X minutes
         - Only get updated maximum once per reconcile as the timestamp already handles that

gkrenn replied:
You are right, the creation does not need to be in the injection reconciler - Added it in the injection reconciler, as I was not sure if the oneagent reconciler will always run before the injection one. But guess this won't change that likely in the future. Will remove it from there. 

The Status is way more elegant did not think about it, thx for pointing out 👍

0sewa0 commented on pkg/controllers/dynakube/oneagent/oneagent_reconciler.go:
```diff
@@ -107,6 +115,15 @@ func (r *Reconciler) Reconcile(ctx context.Context) error {
 		return r.cleanUp(ctx)
 	}
 
+	log.Info("creating Pull Secret")
+
+	err = dtpullsecret.NewReconciler(r.client, r.apiReader, r.dynakube, r.tokens).Reconcile(ctx)
+	if err != nil {
+		log.Error(err, "could not reconcile pull secret")
+
+		return err
+	}
+
```
what if you don't want to deploy OneAgents but only an ActiveGate ? 🙃

gkrenn replied:
That's indeed a problem, still have to add the status condition to the pull secret reconciler and write unit tests for it - And then call it a second time. 

Will do that after my vacation, as I don't want to code in another bug in a hurry now.

gkrenn commented:
Updated the pull secret reconciler and added conditions + cleanup logic

Pull Secret Reconciler is not called from within oneagent reconciler and active gate reconciler

0sewa0 commented on pkg/controllers/csi/provisioner/controller.go:
```diff
@@ -161,7 +161,7 @@ func (provisioner *OneAgentProvisioner) Reconcile(ctx context.Context, request r
 
 func (provisioner *OneAgentProvisioner) setupFileSystem(dk *dynatracev1beta2.DynaKube) error {
-	tenantUUID, err := dk.TenantUUIDFromApiUrl()
+	tenantUUID, err := dk.TenantUUID()
```
this will break the "resilience" feature, because
```
	err = provisioner.setupFileSystem(dk) // <= BOOM, because no tenant-uuidm because network is dead
	if err != nil {
		return reconcile.Result{}, err
	}
	tenantConfig, err := provisioner.setupTenantConfig(dk) // needed for the CSI-resilience feature
	if err != nil {
		return reconcile.Result{}, err
	}
```

gkrenn replied:
Set it back to the old method + follow up ticket created

0sewa0 commented on pkg/controllers/dynakube/dtpullsecret/reconciler.go:
```diff
@@ -18,27 +21,49 @@ const (
+
+		err := r.deletePullSecretIfExists(ctx)
+		if err != nil {
+			log.Error(err, "failed to clean-up OneAgent tenant-secret")
```
i think this log line is not correct 😅 

also if you bubble up the error then don't log it, as it will be logged higher up, and we don't want duplicate error logs

gkrenn replied:
We should just log the error and ignore it in the cleanup, as you described in the active gate connection reconciler. 

Will update wrong log line and remove the return so that the conditions are cleanup up as well like in the active gate connection info reconciler.

0sewa0 commented on pkg/controllers/dynakube/oneagent/oneagent_reconciler.go:
```diff
@@ -107,6 +115,15 @@ func (r *Reconciler) Reconcile(ctx context.Context) error {
 		return r.cleanUp(ctx)
 	}
 
+	log.Info("creating Pull Secret")
```
do we need this log line here?
(you don't have it in the AG reconciler for the same logic)

0sewa0 commented on pkg/controllers/dynakube/oneagent/oneagent_reconciler.go:
```diff
@@ -107,6 +115,15 @@ func (r *Reconciler) Reconcile(ctx context.Context) error {
+
+	err = dtpullsecret.NewReconciler(r.client, r.apiReader, r.dynakube, r.tokens).Reconcile(ctx)
+	if err != nil {
+		log.Error(err, "could not reconcile pull secret")
```
if you bubble up the error then don't log it, as it will be logged higher up, and we don't want duplicate error logs

0sewa0 reviewed (changes requested):
Works as expected, but there are some inconsistencies in logging that we should fix

waodim commented on pkg/controllers/csi/provisioner/controller.go:
```diff
@@ -287,7 +287,7 @@ func (provisioner *OneAgentProvisioner) updateAgentInstallation(
 
 func (provisioner *OneAgentProvisioner) handleMetadata(dk *dynatracev1beta2.DynaKube) (*metadata.TenantConfig, error) {
-	tenantUUID, err := dk.TenantUUIDFromApiUrl()
+	tenantUUID, err := dk.TenantUUIDFromApiUrl() // TODO update to use the tenant uuid from the DynaKube status
```
why isnt this updated in this PR?

waodim reviewed (approved):
LGTM
